primitives/utility/src/lib.rs

```diff
@@ -51,3 +61,157 @@ impl<T: UpwardMessageSender, W: WrapVersion> SendXcm for ParentAsUmp<T, W> {
+				let u128_amount: u128 = asset_balance.try_into().map_err(|_| XcmError::Overflow)?;
+				// Construct required payment
+				let required = (Concrete(location.clone()), u128_amount).into();
+				// Substract payment
+				let unused = payment.checked_sub(required).map_err(|_| XcmError::TooExpensive)?;
+				// record weight
+				self.0 = self.0.saturating_add(weight);
+				// record amount
+				self.1 = self.1.saturating_add(asset_balance);
+				// record multilocation and local_asset_id
+				self.2 = Some((location.clone(), local_asset_id));
+				Ok(unused)
+			},
+			_ => return Err(XcmError::TooExpensive),
+		}
+	}
+
+	fn refund_weight(&mut self, weight: Weight) -> Option<MultiAsset> {
+		log::trace!(target: "xcm::weight", "TakeFirstAssetTrader::refund_weight weight: {:?}", weight);
+		let weight = weight.min(self.0);
+		if let Some((asset_location, local_asset_id)) = self.2.clone() {
+			// Calculate asset_balance
+			// This read should have already be cached in buy_weight
+			let asset_balance =
+				FeeCharger::charge_weight_in_fungibles(local_asset_id, weight).ok()?;
+
+			self.0 -= weight;
+			self.1 = self.1.saturating_sub(asset_balance);
+
+			let asset_balance: u128 = asset_balance.saturated_into();
+			if asset_balance > 0 {
+				Some((asset_location, asset_balance).into())
+			} else {
+				None
+			}
+		} else {
+			None
+		}
+	}
+}
+
+impl<
+		AccountId,
+		FeeCharger: ChargeWeightInFungibles<AccountId, Assets>,
+		Matcher: MatchesFungibles<Assets::AssetId, Assets::Balance>,
+		Assets: fungibles::Mutate<AccountId>
+			+ fungibles::Transfer<AccountId>
+			+ fungibles::Balanced<AccountId>,
+		HandleRefund: TakeRevenue,
+	> Drop for TakeFirstAssetTrader<AccountId, FeeCharger, Matcher, Assets, HandleRefund>
+{
+	fn drop(&mut self) {
+		if let Some((location, _)) = &self.2 {
+			let u128_amount: u128 = self.1.try_into().map_err(|_| XcmError::Overflow).unwrap();
+			HandleRefund::take_revenue(MultiAsset {
+				id: location.clone().into(),
+				fun: Fungibility::Fungible(u128_amount),
+			});
+		}
+	}
+}
+
+/// XCM fee depositor to which we implement the TakeRevenue trait
+/// It receives a fungibles::Mutate implemented argument, a matcher to convert MultiAsset into
+/// AssetId and amount, and the fee receiver account
+pub struct XcmFeesToAccount<Assets, Matcher, AccountId, ReceiverAccount>(
+	PhantomData<(Assets, Matcher, AccountId, ReceiverAccount)>,
+);
+impl<
+		Assets: fungibles::Mutate<AccountId>,
+		Matcher: MatchesFungibles<Assets::AssetId, Assets::Balance>,
+		AccountId: Clone,
+		ReceiverAccount: frame_support::traits::Get<Option<AccountId>>,
+	> TakeRevenue for XcmFeesToAccount<Assets, Matcher, AccountId, ReceiverAccount>
+{
+	fn take_revenue(revenue: MultiAsset) {
+		match Matcher::matches_fungibles(&revenue) {
+			Ok((asset_id, amount)) =>
+				if let Some(receiver) = ReceiverAccount::get() {
+					if !amount.is_zero() {
+						let ok = Assets::mint_into(asset_id, &receiver, amount).is_ok();
```

Comment:
I don't understand why this is minting. Besides the native asset, everything uses reserve backed transfers, so the asset must already be there in the sovereign account of the fee-paying `MultiLocation`. Shouldn't this debit the fee payer and credit the author?

Reply:
I used `HandleCredit` before, similar to what it used in `pallet_asset_tx_payment`. We would achieve the same behavior as we have now, but without an event showing that the author has been credited some tokens. I think the event is important to show how fees are divided between block author and destination address, but I am happy to revert back these changes